File: src/Chapter4/NotSoldComponent.java
package Chapter4;

/*                                                                            *
 * @author Gianluca Capone & Davide Sgobba                                    *
 *                                                                            */

/* This class contains all variables that define the component product        *
 * produced by integrated computer firms and the methods that operate over    *
 * these variables                                                            */
class NotSoldComponent {
    
    // VARIABLES
    double mod;
    /* Merit of design of the component product (M-CO_f,t)                    */
    double muProg;
    /* Mean of the distribution of component technical progress (mu-CO_f,t)   */
    double productionCost;
    /* Component production cost (C-CO_f,t)                                   */

    //TECHNICAL VARIABLES & OBJECTS
    ComputerFirm Firm;
    /* Access to the firm                                                     */
    
    // CONSTRUCTOR
    NotSoldComponent(ComputerFirm FIRM) {
        mod            = 0;
        muProg         = 0;
        productionCost = 0;
        Firm           = FIRM;
    }

    /* This method computes the mod of the component based on the level of    *
     * past mod and the level of public knowledge                             */
    void calcMod() {
        
        double zmax = 0;
        /* Equation 14.b                                                      */
        muProg = Math.log(Firm.ComputerMarket.pkCmp[Firm.tId]) * (1 - Firm.ComputerMarket.internalCum)
                      + Math.log(mod) * Firm.ComputerMarket.internalCum;
        for (int i = 1; i <= Firm.numOfDrawsCmp; i++) {
            double z = Math.exp(muProg + Math.sqrt(Firm.ComputerMarket.sdCmp[Firm.tId])
                         * Firm.ComputerMarket.rng.nextGaussian());
            if (z > zmax) {
                zmax = z;
            }
        }
        if (zmax > mod) {
            mod = zmax;
        }
        if (mod > 0) {
            /* Equation 11                                                    */
            productionCost = Firm.ComputerMarket.nuCmp / mod;
        }
    }
    
    /* This method is activated when exit occurs: it resets the most relevant *
     * variables at the product level                                         */
    void exitComponent() {
        muProg  = 0;
        mod            = 0;
        productionCost = 0;
    }
    
}

File: src_py/Chapter4/not_sold_component.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
NotSoldComponent模块 - NotSoldComponent类的Python实现
转换自Java版本的NotSoldComponent.java
"""

import math

"""
@author Gianluca Capone & Davide Sgobba
Python转换

此类包含定义集成计算机公司生产的组件产品的所有变量以及操作这些变量的方法
"""
class NotSoldComponent:
    
    def __init__(self, firm):
        """
        构造函数
        
        Args:
            firm: 公司对象引用
        """
        # 变量
        self.mod = 0.0               # 组件产品的设计优点 (M-CO_f,t)
        self.mu_prog = 0.0           # 组件技术进步分布的均值 (mu-CO_f,t)
        self.production_cost = 0.0    # 组件生产成本 (C-CO_f,t)
        
        # 技术变量和对象
        self.firm = firm             # 访问公司对象
        
    def calc_mod(self):
        """
        根据过去的mod水平和公共知识水平计算组件的mod
        """
        z_max = 0.0
        # 方程14.b
        self.mu_prog = (math.log(self.firm.computer_market.pk_cmp[self.firm.t_id]) * 
                      (1 - self.firm.computer_market.internal_cum) + 
                      math.log(self.mod) * self.firm.computer_market.internal_cum)
        
        for i in range(1, self.firm.num_of_draws_cmp + 1):
            z = math.exp(self.mu_prog + math.sqrt(self.firm.computer_market.sd_cmp[self.firm.t_id]) * 
                       self.firm.computer_market.rng.gauss(0, 1))
            if z > z_max:
                z_max = z
                
        if z_max > self.mod:
            self.mod = z_max
            
        if self.mod > 0:
            # 方程11
            self.production_cost = self.firm.computer_market.nu_cmp / self.mod
            
    def exit_component(self):
        """
        退出时激活的方法：重置产品层面的最相关变量
        """
        self.mu_prog = 0.0
        self.mod = 0.0
        self.production_cost = 0.0 